### PR TITLE
bot-gate: exit 4 when a round finished but nothing attributes it, plus the degraded-forge procedure

### LINKED ISSUE #23: A skill's text in an agent's context can be superseded — I merged a PR on a procedure the on-disk skill had replaced with a tested script
## What happened

I landed douglaz/lnrent#81 following `pr-with-codex-bot-review`'s **manual** merge procedure —
`gh api .../reactions`, `gh api .../pulls/N/comments`, judge by hand. That procedure is gone. The
on-disk `SKILL.md` now points at **`scripts/bot-gate`** in five places and states that the merge
requires `bot-gate <PR>` exit 0 across six conditions.

I never ran it. The copy of the skill in my context was an earlier revision, loaded long before, and
carried through a context summarisation. Nothing in it mentioned `bot-gate`, so there was nothing to
prompt me to look.

This is worse than reading a stale doc, because `bot-gate`'s own header records that this exact logic
*"was written three times as prose and was wrong three different ways"* — an unenforcing version, an
unsatisfiable version, and one that broke past 100 comments. Following remembered prose is following
one of those known-wrong versions.

## What it cost

`bot-gate` condition 1 is *a wrapper exists whose `Reviewed commit:` equals the tip*. On #81 the API
showed — and still shows — one codex wrapper naming `2496bd20c2`, the **parent** of the merged head
`dab9d55`. I merged on a `+1` reaction plus a human-pasted approval message, which the skill
explicitly says is not what advances the decision. The gate would have blocked; I didn't run it.

## Why the existing note doesn't cover it

`drive/SKILL.md:103-115` warns about two *installs on disk* where the first match may be stale. That
is a different failure: mine was a stale copy **in the model's context**, with a single correct
install on disk the whole time.

## Suggestions

1. **Make executable gates load-bearing in the description**, not just the body. A skill's
   `description` is what survives into a listing; if it names the script ("gate merges with
   `scripts/bot-gate`"), an agent working from a summarised context still sees that a script exists.
2. **A short "verify before you follow" line at the top of any skill that ships a script** — e.g.
   *"If you are recalling this procedure rather than reading it now, re-read the on-disk SKILL.md
   first; the manual steps it used to document have been replaced by `scripts/`."*
3. Consider having the scripts refuse to be bypassed silently — e.g. the skill instructing that a
   merge report must quote `bot-gate`'s output, so an agent that skipped it cannot produce the
   required evidence.

Filed from the operator side: I'd rather the skill make this unmissable than rely on the next agent
happening to have a fresh copy.

### LINKED ISSUE #24: No sanctioned procedure when the forge is degraded: bot-gate is unsatisfiable during a GitHub outage and nothing says what to do instead
## The gap

`bot-gate`'s condition 1 requires *a wrapper whose `Reviewed commit:` equals the tip*. During a
GitHub incident the API can stop serving new review wrappers and stop reporting check runs entirely,
so that condition cannot be met **no matter how long you wait** — not because the bot disapproves,
but because the forge isn't answering.

Grepping `pr-with-codex-bot-review` and `drive` for outage / degraded / "run CI locally" guidance
returns nothing for this case. `drive-status` handles a degraded `gh` for *status reporting*, but no
skill says what an agent should do about **merging** when the forge is impaired.

## The concrete instance

On douglaz/lnrent#81, mid-incident:

- GitHub's checks rollup showed **only** CodeRabbit — the `lint + test` and `web e2e` jobs never
  appeared at all.
- The codex review wrapper for the merged head `dab9d55` never became visible; the API showed (and
  still shows) only one naming the parent `2496bd20c2`.
- A `+1` reaction did appear, timestamped 4 minutes after the commit.

I ran **all seven CI steps locally** on the exact head — clippy in both feature configs, tests
867/819, the fedimint-absence guard, wasm32 clippy+build, and both web-e2e legs — and merged on that
plus the reaction and a human-pasted approval naming the tip.

That may well have been the right call. But I made it ad hoc, and the skill's own position is that a
reaction is *not* what advances the decision. An agent shouldn't be improvising a money-path merge
policy during an outage.

## What would help

1. **Say whether locally-run CI is ever an acceptable substitute**, and if so what must be recorded:
   the exact commands, the head SHA they ran against, and that they were run on the *merged* head
   rather than an earlier one.
2. **Say whether a human-relayed bot verdict counts.** A pasted "Reviewed commit: <sha>" is evidence
   an agent cannot verify. Either it's admissible with the SHA cross-checked against the local head
   (which is checkable), or it isn't and the merge waits.
3. **Distinguish "the bot has not reviewed" from "the forge cannot tell me whether it reviewed."**
   Right now both look like a missing wrapper, and only the first is a reason to keep waiting.
4. Possibly a `bot-gate` exit code for *forge-degraded* rather than a plain BLOCK, so the caller can
   tell an unmet condition from an unanswerable one.

Related: #23 (a stale skill copy in context meant I ran the superseded manual procedure here rather
than `bot-gate` at all).

### PR DESCRIPTION
**Pull request by douglaz**
Closes #23. Closes #24.

Two failures from the same day, both the same shape: **an agent that could not tell "there is no signal" from "I am not being told."**

## #24 — the gate was unsatisfiable during an outage

`bot-gate`'s first condition is *a wrapper naming the tip*. During a GitHub incident that cannot be met **no matter how long you wait** — the API keeps answering `200` while new wrappers stop appearing. A caller polling for one waits forever on a condition the forge will never produce.

**New verdict `BLOCKED_UNATTRIBUTED`, exit 4.**

| Code | Verdict | Means |
|---|---|---|
| 0 | `NO_PENDING_EVIDENCE` | not clearance |
| 1 | `BLOCKED` | a conjunct is unmet; usually **wait** |
| 4 | `BLOCKED_UNATTRIBUTED` | a round finished; nothing says which tree it read |

It fires only when every other conjunct already holds, no wrapper names the tip, and the bot's `+1` **post-dates the tip's commit**. Per the bot's own about-box — *"if Codex has suggestions, it will comment; otherwise it will react with 👍"* — a 👍 is a *completed* round, attached to the PR and never to a commit. A 👍 older than the tip provably belongs to an earlier tree and leaves a plain exit 1.

**Exit 4 refuses the merge exactly as 1 does.** Any caller doing `bot-gate <PR> && gh pr merge` is unaffected. It exists for a caller that *loops*: exit 1 means wait, exit 4 means waiting cannot help, because the signal that would attribute the round is the missing one.

### Two things I deliberately did not claim

**It has never met live traffic.** The branch is derived from the bot's *documented contract*, not from an observation here — this repo's own 19-round sample on PR #16 saw the `+1` fire **zero times**. The skill says so rather than implying it was measured.

**bot-gate still cannot detect an outage.** An incident that hides a wrapper is indistinguishable, from the API's own responses, from a bot that never reviewed. Inventing a detector would be exactly the over-claim ADR 0004 exists to prevent. So § 8b sends you to the *status page* — a missing wrapper is not evidence of an outage; a degraded component, or CI jobs absent from a rollup that should list them, is.

### § 8b, the degraded-forge procedure

Answers the three questions the issue asked:

- **Locally-run CI can substitute — for CI.** It is stronger than a green check mark because you watched it. What makes it admissible is the record: exact commands with real exit codes, **the head SHA they ran against** (a local run on an earlier head proves nothing about the tree that lands), and which forge jobs went unrun. It does **not** substitute for the review — no local command tells you what a reviewer would have said.
- **A human-relayed verdict is admissible only for the checkable part.** A pasted `Reviewed commit: <sha>` is a claim about a message you cannot see, but the SHA is verifiable against local HEAD and a mismatch settles it. Cross-check it and say you did.
- **Record it as a degraded merge.** The failure this prevents is not merging during an outage — that is sometimes correct — it is improvising a merge policy during one and leaving no trace of the trade.

## #23 — the skill in context can be superseded

A PR merged on the *manual* reaction-and-comment procedure that `scripts/bot-gate` replaced. The copy in context predated the script, so nothing prompted a look. The gate would have blocked that merge: the only wrapper the API held named the **parent** of the merged head.

- **`description` now states the merge decision is `bot-gate`'s, not judgment's** — descriptions survive into listings, so an agent working from a summarised context still learns a script exists.
- **A verify-before-you-follow block** at the top of both script-shipping skills. The tell it names: *a procedure you can recite in full without having opened a file this turn.*
- **§ 8a — the merge report must QUOTE the gate's verdict, tip and window.** This is the countermeasure that actually bites: it is the one step an agent who skipped the gate cannot produce without lying outright. Every other step leaves an artifact that could be reconstructed after the fact.

## Verification

9 new fixtures, **113 pass / 0 fail** (was 104). Each shown red against its own fix rather than merely passing:

| Mutation | Fixtures that went red |
|---|---|
| delete the `BLOCKED_UNATTRIBUTED` branch | exit-4 case, JSON verdict |
| drop the tip-timestamp bound | 👍 older than the tip |
| admit undated reactions | undated 👍 |
| drop the other conjuncts from the `elif` | unresolved threads, pending review |
| escalate the reactions failure to exit 3 | both query-failure cases |

Tree restored and re-verified at 113/113 after each. `drive-status.test` 70/70.

The fixtures use a fixed far-future literal rather than `$NOW_UTC`: the fixture repo's commit is created *after* `NOW_UTC` is sampled, so "now" can land a second before the tip and the case would pass or fail on the clock.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by douglaz** [comment]
All three P2s addressed in 882fd1a. Two were real defects; the third I had already fixed in this branch before the review landed — independently, which is a useful corroboration.

**FIXED — Treat a later `eyes` reaction as an active round.** Real, and the reasoning is exactly right: `ROUND_ACTIVE` is gated on `WRAPPER_AT`, which is empty on *every* path that reaches exit 4, so it was false **by construction rather than by evidence**. An older `+1` with a newer `eyes` therefore produced "waiting may not help" while an `eyes` positively said a round was running — the worst direction for this verdict, since its whole job is to tell a caller when to stop polling. Both reactions are now read from the same response; an `eyes` at or after the `+1` withdraws the claim, and so does one that cannot be placed in time.

**FIXED (wording) — Bound the reaction to when this tip became the PR head.** The failure case is correct and I could not close it. The honest boundary is when the tip became the head, and nothing reports that: the timeline carries `head_ref_force_pushed` but not an ordinary push — the same gap `bot-gate`'s round-start list already documents as a residual. So commit-locally → bot thumbs the old remote head → push does produce a false exit 4, and the `+1` survives because GitHub allows one reaction per user per item.

Kept, with the claim weakened rather than the mechanism removed:

- the failure is **one-directional** — it can only turn a block into a *differently-worded* block, never into a clearance
- § 8b's first instruction, confirm a real incident on the status page, is precisely what disproves it
- a tighter bound needs a timestamp the forge does not expose, and inventing one is the over-claim ADR 0004 exists to prevent

So "waiting cannot help" became "waiting **may** not help", every site now says try `@codex review` **first**, and the residual is written down in the script, the skill, and `drive`'s LAND notes.

**ALREADY FIXED — Parse Git's numeric timezone on BSD date.** Landed as `%ct` before this review arrived. One refinement to the finding's premise: `%cI` emits `Z` under a *UTC* committer and `-07:00` otherwise, so the break is TZ-dependent rather than unconditional on macOS — which makes it worse, not better, since a macOS clone with UTC commits would work and hide it.

Worth flagging what the suite does **not** do here. I ran the reversion as a mutation and it **stays green** on this box — `date -u -d` accepts both forms, verified including under `TZ=America/Los_Angeles`. My first attempt at that mutation *did* go red, but only because it half-applied; it was failing for the wrong reason. Both files now say so outright: the three boundary fixtures pin the **comparison**, and the format choice is held by review only. A test that cannot fail against the defect it names is not covering it.

Suite: **119 pass / 0 fail** (104 before this branch). New fixtures mutation-proved — deleting the `eyes` guard reddens the two `eyes` cases, `>=`→`>` reddens the at-the-boundary case. `drive-status.test` 70/70.

File: skills/pr-with-codex-bot-review/scripts/bot-gate [hunk 1]
```diff
@@ -844,6 +844,56 @@ PR_HEAD_FINAL=$(gh pr view "$PR" -R "$REPO" --json headRefOid -q .headRefOid 2>/
...
+     && printf '%s' "$_REACT" | jq -e . >/dev/null 2>&1; then
+    # Undated reactions are dropped rather than treated as "now": an undatable signal
+    # cannot be shown to post-date the tip, and this branch only ever fires on evidence.
+    _THUMBS_AT=$(printf '%s' "$_REACT" | jq -r --arg bot "$BOT_LOGIN" '
+      [.[][] | select(.user.login==$bot) | select(.content=="+1")
+       | select(((.created_at // "") | length) > 0) | .created_at] | max // ""' 2>/dev/null) \
+      || _THUMBS_AT=""
+    if [[ -n "$_THUMBS_AT" && -n "$_TIP_EPOCH" ]]; then
+      _THUMBS_EPOCH=$(iso_to_epoch "$_THUMBS_AT") || _THUMBS_EPOCH=""
+      [[ -n "$_THUMBS_EPOCH" ]] && (( _THUMBS_EPOCH >= _TIP_EPOCH )) && THUMBS_UNATTRIBUTED=true
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/pr-with-codex-bot-review/scripts/bot-gate:921]: Addressed in 882fd1a by weakening the claim rather than the mechanism — the failure case is correct and I could not close it. The honest boundary is when the tip became the PR head, and nothing reports that: the timeline carries `head_ref_force_pushed` but not an ordinary push, the same gap bot-gate's round-start list already documents. Kept because the failure is one-directional (it can only turn a block into a differently-worded block, never a clearance) and § 8b's first instruction — confirm a real incident on the status page — is exactly what disproves it. A tighter bound needs a timestamp the forge does not expose, and inventing one is the over-claim ADR 0004 exists to prevent. So: "waiting cannot help" became "may not help", every site says try `@codex review` first, and the residual is written down in the script, the skill, and drive's LAND notes.

File: skills/pr-with-codex-bot-review/scripts/bot-gate [hunk 2]
```diff
@@ -844,6 +844,56 @@ PR_HEAD_FINAL=$(gh pr view "$PR" -R "$REPO" --json headRefOid -q .headRefOid 2>/
...
+  # latter, and it is the rewritten commit that got pushed.
+  _TIP_AT=$(git show -s --format=%cI "$TIP" 2>/dev/null) || _TIP_AT=""
+  _TIP_EPOCH=$(iso_to_epoch "$_TIP_AT") || _TIP_EPOCH=""
+  if _REACT=$(gh api --paginate --slurp "repos/$REPO/issues/$PR/reactions" 2>/dev/null) \
+     && printf '%s' "$_REACT" | jq -e . >/dev/null 2>&1; then
+    # Undated reactions are dropped rather than treated as "now": an undatable signal
+    # cannot be shown to post-date the tip, and this branch only ever fires on evidence.
+    _THUMBS_AT=$(printf '%s' "$_REACT" | jq -r --arg bot "$BOT_LOGIN" '
+      [.[][] | select(.user.login==$bot) | select(.content=="+1")
+       | select(((.created_at // "") | length) > 0) | .created_at] | max // ""' 2>/dev/null) \
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/pr-with-codex-bot-review/scripts/bot-gate:917]: Fixed in 882fd1a. Real, and the diagnosis is exactly right: ROUND_ACTIVE is gated on WRAPPER_AT, which is empty on every path reaching exit 4, so it was false **by construction rather than by evidence**. The gate would have said "waiting may not help" while an `eyes` positively said a round was running — the worst direction for a verdict whose whole job is telling a caller when to stop polling. Both reactions now come from the same response; an `eyes` at or after the `+1` withdraws the claim, and so does one that cannot be placed in time. Two fixtures, mutation-proved.

File: skills/pr-with-codex-bot-review/scripts/bot-gate [hunk 3]
```diff
@@ -844,6 +844,56 @@ PR_HEAD_FINAL=$(gh pr view "$PR" -R "$REPO" --json headRefOid -q .headRefOid 2>/
...
+# refuse a decidable PR over a signal that can only ever make the answer more specific,
+# never more permissive.
+THUMBS_UNATTRIBUTED=false
+if [[ "$WRAPPER_TIP" -eq 0 ]]; then
+  # The tip's own commit date bounds what the reaction could be about: a `+1` older than
+  # the commit provably belongs to an earlier tree and leaves a plain BLOCK. Committer
+  # date, not author date — a rebase or amend rewrites the former and preserves the
+  # latter, and it is the rewritten commit that got pushed.
+  _TIP_AT=$(git show -s --format=%cI "$TIP" 2>/dev/null) || _TIP_AT=""
+  _TIP_EPOCH=$(iso_to_epoch "$_TIP_AT") || _TIP_EPOCH=""
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/pr-with-codex-bot-review/scripts/bot-gate]: Already fixed in this branch as `%ct` before the review landed — good independent corroboration. One refinement to the premise: `%cI` emits `Z` under a UTC committer and `-07:00` otherwise, so the break is TZ-dependent rather than unconditional on macOS — which makes it worse, since a macOS clone with UTC commits would work and hide it. Also worth recording: I ran the reversion as a mutation and the suite **stays green** here, because GNU `date -u -d` accepts both forms (verified under TZ=America/Los_Angeles too). My first attempt at that mutation went red only because it half-applied — failing for the wrong reason. Both files now say the boundary fixtures pin the comparison and that this format choice is held by review only.
